hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java

```diff
@@ -0,0 +1,319 @@
+                    } catch (Throwable e) {
+                        throw new AssertionError("For [" + fieldType + ", " + lowerBoundType + ", " + upperBoundType + "]: " + e, e);
+                    }
+                }
+            }
+        }
+        assertEquals(TESTED_TYPES.length * TESTED_TYPES.length * TESTED_TYPES.length, cycleCounter);
+    }
+
+    @Test
+    @Ignore(value = "Un-ignore after engines merge")
+    public void rowNumericBetweenPredicateTest() {
+        putAll(0, 1, 5, 10, 15, 25, 30);
+        checkValues("SELECT * FROM map WHERE this BETWEEN ROW(0, 0) AND ROW(3, 10)", INTEGER, new Integer[][]{
+                new Integer[]{0, 0}, new Integer[]{1, 1}, new Integer[]{2, 5}, new Integer[]{3, 10},
+        });
+    }
+
+    protected void checkValues(
+            String sql,
+            SqlColumnType expectedType,
+            Object[] expectedResults,
+            Object... params
+    ) {
+        List<SqlRow> rows = execute(member, sql, params);
+        assertEquals(expectedResults.length, rows.size());
+        if (rows.size() == 0) {
+            return;
+        }
+
+        if (rows.get(0).getObject(0) instanceof Integer) {
+            rows.sort(Comparator.comparingInt(a -> a.getObject(0)));
+        } else if (rows.get(0).getObject(0) instanceof Comparable) {
+            rows.sort((a, b) -> ((Comparable<?>) a.getObject(0)).compareTo(b.getObject(0)));
+        }
+
+        for (int i = 0; i < expectedResults.length; i++) {
+            SqlRow row = rows.get(i);
+            assertEquals(expectedType, row.getMetadata().getColumn(0).getType());
+            assertEquals(expectedResults[i], row.getObject(0));
+        }
+    }
+
+    protected void checkSuccessOrFailure(
+            String sql,
+            Tuple2<List<SqlRow>, HazelcastSqlException> expectedOutcome,
+            Object... params
+    ) {
+        try {
+            List<SqlRow> rows = execute(member, sql, params);
+            assertNull(expectedOutcome.f1());
+            assertEquals(expectedOutcome.f0().size(), rows.size());
+            List<SqlRow> expectedResultsList = expectedOutcome.f0();
+            for (int i = 0; i < rows.size(); i++) {
+                Object actualObject = rows.get(i).getObject(0);
+                Object expectedObject = expectedResultsList.get(i).getObject(0);
+                assertEquals(expectedObject, actualObject);
+
+                SqlColumnType expectedType = expectedResultsList.get(i).getMetadata().getColumn(0).getType();
+                SqlColumnType actualType = rows.get(i).getMetadata().getColumn(0).getType();
+                assertEquals(expectedType, actualType);
+            }
+        } catch (HazelcastSqlException e) {
+            assertNotNull(expectedOutcome.f1());
+            // Expected : ... Parameter at position 0 must be of BIGINT type, but VARCHAR was found
+            // Actual   : ... Parameter at position 0 must be of TINYINT type, but VARCHAR was found
```

Comment:
This actually is an issue, it has to say "BIGINT" in both cases. I added a specific test just for this. Basically we need to replicate [this behavior](https://github.com/hazelcast/hazelcast/blob/master/hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/BinaryOperatorOperandTypeInference.java#L55-L66) of binary operator.

Reply:
Done.